hal/device.h

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (C) 2019-2020 BiiLabs Co., Ltd. and Contributors
+ * All Rights Reserved.
+ * This is free software; you can redistribute it and/or modify it under the
+ * terms of the MIT license. A copy of the license can be found in the file
+ * "LICENSE" at the root of this distribution.
+ */
+
+#ifndef HAL_DEVICE_H
+#define HAL_DEVICE_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "defined_error.h"
+
+/*! device initialization entry point */
+#define DECLARE_DEVICE(name)                                  \
+  void init_##name##_init(void) __attribute__((constructor)); \
+  void init_##name##_init(void) { name##_device_type.op->init(); }
+
+typedef struct device_type device_t;
+
+struct device_operations {
+  retcode_t (*init)(void);            /**< initialize for device */
+  void (*fini)(void);                 /**< destructor for device  */
+  retcode_t (*get_key)(uint8_t *);    /**< get device private key */
+  retcode_t (*get_device_id)(char *); /**< get device id          */
+};
+
+struct uart_operations {
+  retcode_t (*init)(const char *device);        /**< init uart */
+  void (*write)(const int fd, const char *cmd); /**< write command to uart */
+  char *(*read)(const int fd);                  /**< read from uart */
+  void (*clean)(const int fd);                  /**< flush uart buffer */
+};
+
+struct secure_store_operations {
+  retcode_t (*init)(void); /**< init secure storage */
+  /**
+   * @brief Write an item to secure storage
+   *
+   * @param[in] name Name of item
+   * @param[in] buf Pointer to data
+   * @param[in] buf_size Length of data
+   *
+   * @return
+   * - #RET_OK on success
+   * - #NO_MEMORY on no memory error
+   * - #UNAVAILABLE on secure storage is currently unavailable
```

Comment:
I think the grammar is a little weird.
Change to `unavailable secure storage` for simplicity?